mod.go

```diff
@@ -0,0 +1,669 @@
+// uint256: Fixed size 256-bit math library
+// Copyright 2021 uint256 Authors
+// SPDX-License-Identifier: BSD-3-Clause
+
+package uint256
+
+import (
+	"math/bits"
+	"sync"
+)
+
+// Cache for reciprocal()
+//
+// Each cache set contains its own mutex to reduce lock contention in highly
+// multithreaded settings.
+//
+// Adjust cacheIndexBits and cacheWays to scale the size of the cache.
+// Reasonable values are quite small, e.g. cacheIndexBits from 2 to 10, and
+// cacheWays from 3 to 11 (3+8n makes set size an integer number of cachelines)
+
+const (
+	cacheIndexBits = 9
+	cacheWays = 3
+
+	cacheSets = 1 << cacheIndexBits
+	cacheMask = cacheSets - 1
+)
+
+type cacheSet struct {
+	hit	uint64
+	miss	uint64
+
+	rw	sync.RWMutex
+	mod	[cacheWays]Int
+	inv	[cacheWays][5]uint64
+}
+
+var cache [cacheSets]cacheSet
+
+func cacheStats() (hit, miss uint64) {
+	for s:=0; s<cacheSets; s++ {
+		hit  += cache[s].hit
+		miss += cache[s].miss
+	}
+	return
+}
+
+// Some utility functions
+
+func add256(x, y Int, c uint64) (Int, uint64) {
+	var z Int
+	z[0], c = bits.Add64(x[0], y[0], c)
+	z[1], c = bits.Add64(x[1], y[1], c)
+	z[2], c = bits.Add64(x[2], y[2], c)
+	z[3], c = bits.Add64(x[3], y[3], c)
+	return z, c
+}
```

Comment:
We already have 
```
func (z *Int) Add(x, y *Int) *Int {
	var carry uint64
	z[0], carry = bits.Add64(x[0], y[0], 0)
	z[1], carry = bits.Add64(x[1], y[1], carry)
	z[2], carry = bits.Add64(x[2], y[2], carry)
	z[3], _ = bits.Add64(x[3], y[3], carry)
	return z
}
```
and
```
// AddOverflow sets z to the sum x+y, and returns z and whether overflow occurred
func (z *Int) AddOverflow(x, y *Int) (*Int, bool) {
	var carry uint64
	z[0], carry = bits.Add64(x[0], y[0], 0)
	z[1], carry = bits.Add64(x[1], y[1], carry)
	z[2], carry = bits.Add64(x[2], y[2], carry)
	z[3], carry = bits.Add64(x[3], y[3], carry)
	return z, carry != 0
}
```
Seems like we already have the same code repeated twice, I'd like it if we don't repeat it a third time. We should be able to combine them, somehow

Reply:
You are right, and I have just removed the new ones.

Reply:
Ack. Third variant of addition (and subtraction) deleted.